### DISCUSSION TITLE
Performance is substantially slow on PyPy

### DISCUSSION THREAD

**Discussion by om26er** [psycopg/psycopg#539] [Ideas]
My preliminary tests suggest psycopg on CPython 3.10 is at least 100% faster than PyPy 3.9 -- Is that expected ? -- Also what would be the right place to start if we wanted to fix that ?

**Comment by dvarrazzo** [comment]
Have you tested Python only, or psycopg[c]? I would find likely that PyPy is 100% slower than psycopg[c], which is some 20x faster than the python implementation. If PyPy is 100% slower than Python that'd be surprising.

Considering that I've never run psycopg on pypy, and that the C code in psycopg-c is finely optimized, at memcopy and refcount level, I'd say yes, I would have expected it even slower actually.

I also have no idea how PyPy talks to the libpq: my guess is via ctypes, which I understand being much slower than cffi, and I don't believe anyone has written a cffi libpq wrapper without telling us.

I have zero experience in optimizing PyPy. Two possible strategies are:

- profiling
- asking @antocuni

Also, does your preliminary test involve a busy loop of fast queries on localhost? Or have you run a test on a server connected to the client via a fast network? I have learned [the hard way](https://github.com/psycopg/psycopg/issues/448#issuecomment-1368491408) that tests on localhost hardly representative of the reality.

**Comment by cirospaciari** [comment]
PyPy + CFFI module is really performant but if it tries to execute a native C extension written for CPython will be really really slow, take a look at https://www.pypy.org/posts/2018/09/inside-cpyext-why-emulating-cpython-c-8083064623681286567.html

The future for C in PyPy is https://hpyproject.org/ with makes it compatible with CPython and GraalPython

@dvarrazzo I will try to take a look at psycopg performance on PyPy and also try to contribute a little bit.

**Comment by om26er** [comment]
My test has been on a busy loop, yes. I am not using Psycopg directly, rather through SQLAlchemy (and asyncio). The simple test for me has been to just get a sqlalchemy model in a loop.

@cirospaciari I should be able to help as well but I believe your understanding of PyPy/CFFI is much better than mine

**Comment by cirospaciari** [comment]
> My test has been on a busy loop, yes. I am not using Psycopg directly, rather through SQLAlchemy (and asyncio). The simple test for me has been to just get a sqlalchemy model in a loop.
> 
> @cirospaciari I should be able to help as well but I believe your understanding of PyPy/CFFI is much better than mine

I think we can try a CFFI or HPy version, I need to read more about this library but with some time we can convert it.
Pure Python version + optimized parts in CFFI /HPy should give us a better performance on PyPy. (PyPy JIT's pure Python, but Python will always have overhead over pure C), CFFI has a lot of overhead so we need to be careful.

I'm working on socketify using CFFI for a while https://github.com/cirospaciari/socketify.py so i have some experience optimizing for PyPy
